### PR TITLE
feat: add Runtime interface

### PR DESCRIPTION
**Pull request by andrewrynhard**
Instead of passing around a struct, it is better if we pass around an
interface that describes the behavior we want. The Runtime interface
provides a common place to describe runtime specific parameters. This
initial implementation offers the runtime mode, the platform specifics,
and the config.

### CODE REVIEW

File: internal/app/machined/internal/phase/phase.go [hunk 1]
```diff
@@ -52,22 +52,27 @@ func NewRunner(config config.Configurator) (*Runner, error) {
...
 
-	mode := platform.Mode()
-	switch mode {
-	case runtime.Metal:
-		fallthrough
-	case runtime.Cloud:
+	var mode runtime.Mode
+
+	switch platform.Name() {
+	case "Container":
```
Comment (andrewrynhard) [inline internal/app/machined/internal/phase/phase.go]: Trying to find a better way to handle this.

**Comment by andrewrynhard** [comment]
Follow up to this will be to allow for runtime errors so that we can avoid panics.

File: internal/pkg/runtime/runtime.go [hunk 2]
```diff
@@ -22,3 +27,41 @@ const (
...
+func NewRuntime(m Mode, p platform.Platform, c config.Configurator) Runtime {
+	return &DefaultRuntime{
+		m: m,
+		p: p,
+		c: c,
+	}
+}
+
+// DefaultRuntime implements the Runtime interface.
+type DefaultRuntime struct {
```
Comment (bradbeam) [inline internal/pkg/runtime/runtime.go]: feel like this is another instance of naming is hard. `Default` doesn't seem like the right fit, but at the same time i don't think these implementation details leak through at all, so not sure it really matters much.

File: internal/app/machined/internal/phase/phase.go (hunk 1 above)
Comment (bradbeam) [reply internal/app/machined/internal/phase/phase.go]: Why move it from `platform.Mode()`?

File: internal/pkg/runtime/runtime.go (hunk 2 above)
Comment (andrewrynhard) [reply internal/pkg/runtime/runtime.go]: Yep. Not sure what else to call it.

File: internal/app/machined/internal/phase/phase.go (hunk 1 above)
Comment (andrewrynhard) [reply internal/app/machined/internal/phase/phase.go]: Import cycle issue. Still chewing on this to find a better way. But the issue is that `runtime` imports platform, and platform imports `runtime`.

**Comment by andrewrynhard** [comment]
This is the first PR in a series aimed at standardizing our boot process.

**Pull request merged**